### ISSUE TITLE
Create entry points to web client

### ISSUE THREAD

**Issue by wabain** [DDMAL/misirlou#25]
Currently the HTML client app just consists of a `/search` page, an upload page, and the various manifests. We need to add an entry point. @ahankinson suggested moving `/search` to the root URL - I'm envisioning this as having a search input at `/` which takes you to `/search?q=whatever` when you type. That also dovetails nicely with the existing JSON API, which doesn't accept `/search` without a query.

@fujinaga has also requested that we add a list of sample manifests. This could go in the header, and another natural place to integrate it would be as a prompt when the user hasn't entered a search term - something like:

> Try searching for **st. gallen** [or something] or [browse notable manuscripts](#)

**Comment by ahankinson** [comment]
Can we not do 'example.com/?q=blahblah'? I would prefer this to redirecting to 'search' for the results.

Browsing manifests might be a good thing to start with. Check out http://digital.bodleian.ox.ac.uk, esp. as you scroll down. Maybe something like this?

**Comment by wabain** [comment]
Making that switch probably shouldn't be a problem in terms of implementation (cc @lexpar). I think it would take about two lines of code on the client-side. I like the unfolding effect on the Digital Bodleian page!

What would requesting `/` as JSON without a search query return after this change? Should it be the API docs still, or something else? With the current search view, I think it would return a 400. Presumably that's not what we'd want.

**Comment by agpar** [comment]
Is the 'switch' you're referring to making the root an html endpoint which will redirect to `/search` if a `?q=` is provided?

**Comment by wabain** [comment]
The switch, as I understand it, would be making the root the JSON endpoint for search, so you'd have `/?q=...` instead of `/search/?q=...`. The HTML would correspond.

**Comment by agpar** [comment]
That's the opposite of what Andrew said he wanted. No? Or, its the opposite of what he said he didn't want. But that's confusing.

**Comment by wabain** [comment]
I'm reading "Can we not do 'example.com/?q=blahblah'?" as "Is there a reason we could not do it, because doing that would be my preference". I think you read it differently?

**Comment by agpar** [comment]
@ahankinson help us out.

**Comment by ahankinson** [comment]
`example.com/?q=blah`

Since search is the main function of this app, so it makes sense to do everything from the root.

**Comment by wabain** [comment]
So for the HTML client the behaviour would be:
- At `/` there's a search input and below it there are links to manifests, laid out similar to the Digital Bodleian homepage.
- Typing in the search input brings up search results, with the associated JSON endpoint being `/?q=...`. If I stuck pretty closely to the current search implementation then I guess the manifest links would be replaced by search results as soon as the user began typing, and the manifest links would reappear if the search input is cleared?

In terms of JSON, I think the only question would be what would happen at the root URL with no query parameter. Having it be API docs strikes me as confusing, but I'm not sure we would want it to return a 400 either.

**Comment by ahankinson** [comment]
I think for the JSON the appropriate thing is to return an empty results list, with perhaps another section that gives the path for the 'upload' address so that other clients _could_ attempt to configure themselves and add manifests.

I'm sure you could also use the server-reported uploads path in your client too.

**Comment by ahankinson** [comment]
You might have the results list 'push' the manifest lists down, and then 'slide' up when the search input is cleared.

This is web 2.0 you know. We're not barbarians. :)

**Comment by ahankinson** [comment]
Or is it 3.0 now? I've lost count.

**Comment by wabain** [comment]
It's probably like 35.0.3 because of semantic versioning. Pushing the results down sounds good though. I'll do that.

**Comment by ahankinson** [comment]
Darn Chrome ruined illogical and arcane versioning schemes for everyone.

**Comment by agpar** [comment]
I cooked up this functionality on 78d972022a4731b7b498769c96e43a31839e2ea1 on a dev branch so you both could put some eyes on it and answer a couple questions I have.
1. Should we still have a `/search` endpoint? My feeling is yes. I like having that distinction.
2. What kind of manifest links should get served up on `/`? It looks like on http://digital.bodleian.ox.ac.uk/ they've hardcoded in some interesting and colourful ones. Do we want to do this, or maybe just the 10 most recent manifests, with a link to see more? 
3. Is `routes` the appropriate key for a dict of endpoints to be served on `/`? I just grabbed this name from what Rodan was doing, but maybe one of you can confirm that this is good or wants something else.

**Comment by ahankinson** [comment]
1. `http://www.google.com/search/` = 404. I'm looking at this like a search engine, so 'search' is the primary purpose of it, not a sub-purpose. I'm good with not having a `/search` endpoint.
2. I'm OK with the 10 most recent. Less management and simpler to hard-code.
3. `routes` sounds good to me.

**Comment by agpar** [comment]
OK @wabain, I'm going to put all the functionality of searching on to `/` on to the wip-landing-page branch, but I won't actually remove the search endpoint, as I'm guessing it would break things for you. That should let you move things over, then we can nix the `/search` endpoint.

The only change that could effect what you were working on on that branch is that `/` now has SinglePageAppRenderer as it's default renderer. I'll hold off on pushing this until you give the OK.

**Comment by wabain** [comment]
@lexpar looks good. I can get to the refactoring later this week.

Re Google's URL structure: the entry point is google.ca but when you enter a query you get taken to https://www.google.ca/search?q=query or whatever. So the logic is a lot like Alex's initial JSON implementation: trying to get `/search` gives you a 400 level response, but that's the endpoint you use to get search results.

**Comment by agpar** [comment]
Okay great. I pushed the changes to wip-landing-page.

As an aside, I figured there wasn't much of a point in heading into the office today, as I didn't think any one would be around. Are there any days over the break that you're planning on going into work @wabain? I think there is a lot of value to us being in the same room at least some amount of time.

**Comment by ahankinson** [comment]
Can you think of a valid technological reason to maintain the search endpoint? I'm willing to concede the point if there is something that is necessary or easier with two endpoints but my naive inclination is that only having one is simpler. Thoughts?

**Comment by agpar** [comment]
My feeling was just that it made sense to have a point that was clearly used to send search requests to. It's not obvious that sending queries into the root should return search results. It is obvious that sending queries into `/search` should return search results.

I honestly don't see a problem with keeping both of them. Even if the `/search` endpoint is never used, I still feel it should be there. Otherwise we are failing to offer a 'most obvious' way to search the site.

**Comment by wabain** [comment]
I'm pretty agnostic about this because there's very little difference in terms of user experience or difficulty of implementation, but one reason to maintain a separate endpoint is because there's a mismatch of resource types. Even if you return an empty result list in the absence of a query, that still means something different than an empty result list given some query; clients need to make a distinction between a result list which is empty because nothing matches and one which is empty because nothing was searched for. Really, the resource that `/` represents is an interface to make search requests, rather than one representing search results. (This reminds me that I never commented on DDMAL/cantus#250. I'll do that when I have a free second.)

@lexpar I agree! However I'll be in Ontario for most of the break. I get back on the first and I'll probably come in sometime after that.

**Comment by wabain** [comment]
I do think that we should only keep one of `/?q=whatever` and `/search?q=whatever` though.

**Comment by agpar** [comment]
Right. I'm fine going either way. The impact is small and my feelings about it are entirely subjective (maybe it _is_ confusing that a search website has two urls that do the exact same thing).

**Comment by ahankinson** [comment]
> Even if you return an empty result list in the absence of a query, that still means something different than an empty result list given some query; clients need to make a distinction between a result list which is empty because nothing matches and one which is empty because nothing was searched for.

This is a very good point. This is usually the behaviour that is implemented for `/search` endpoints; that is, if it's a `GET` return the "search" page with no results, and if it's a `POST` return the search results with a "results" page. The methods inform the intent of the action, rather than the endpoint.

If we were to get pseudo-philosophical for a minute, we should think of whether we think "search" in this case is a verb ("Search for this thing...") or whether search is a noun ("Your search for this thing..."). This will inform the design, since only nouns (i.e., resources) should have endpoints, and verbs should be methods. (cf [proposed HTTP SEARCH method](https://tools.ietf.org/html/draft-snell-search-method-00).)

**Comment by wabain** [comment]
Interesting:

> The SEARCH method is used to initiate a server-side search.  Unlike
>    the HTTP GET method, which requests that a server return a
>    representation of the resource identified by the effective request
>    URI (as defined by [RFC7230]), the SEARCH method is used by a client
>    to ask the server to perform a query operation (described by the
>    request payload) over some set of data scoped to the effective
>    request URI.

So a request would be something like

```
SEARCH /manifests HTTP/1.1
Host: example.org
Content-Type: text/query
Accept: application/json

...
```

I think conceptually what we have now is a search resource (noun), so using GET is fine. Working out the rationale for the SEARCH method (caching, entity-body semantics...) led me down a nice rabbit-hole though.

**Comment by wabain** [comment]
Not sure where we left this. Should I go ahead with moving search handling to `/`?

**Comment by ahankinson** [comment]
Yes

**Comment by wabain** [comment]
FYI I'm pushing some messy work to the wip-landing-page branch. I'm going to get it cleaned up and merged onto master now because the branches are diverging too far.

**Comment by wabain** [comment]
Okay, I've got the updates onto master. Things @lexpar should know about:
- I copied the changes which had been made to the search view to the master view (9febebcb658af37d0a333f625b1c736f89dde1d9). The search view can probably be deleted now.
- I made a `/manifests/recent` endpoint to have something to code against (2b55a464dbafb53e9a5678495b555c5bcdf74f35). It returns 12 most recently created manifests. This might or might not be what we really want to do. Another option would be to order and paginate `/manifests` and use that instead.
